src/widgets/panel/panel.js

```diff
@@ -0,0 +1,116 @@
+import React, { render, unmountComponentAtNode } from 'preact-compat';
+import cx from 'classnames';
+import { getContainerNode, prepareTemplateProps, warn } from '../../lib/utils';
+import { component } from '../../lib/suit';
+import Panel from '../../components/Panel/Panel';
+
+const suit = component('Panel');
+
+const renderer = ({ containerNode, cssClasses, templateProps }) => ({
+  renderingOptions,
+  hidden,
+}) => {
+  let bodyRef = null;
+
+  render(
+    <Panel
+      cssClasses={cssClasses}
+      hidden={hidden}
+      templateProps={templateProps}
+      data={renderingOptions}
+      onRef={ref => (bodyRef = ref)}
+    />,
+    containerNode
+  );
+
+  return { bodyRef };
+};
+
+const usage = `Usage:
+const widgetWithHeaderFooter = panel({
+  [ templates.{header, footer} ],
+  [ hidden ],
+  [ cssClasses.{root, noRefinementRoot, body, header, footer} ],
+})(widget);
+
+const myWidget = widgetWithHeaderFooter(widgetOptions)`;
+
+export default function panel({
+  templates = {},
+  hidden = () => false,
+  cssClasses: userCssClasses = {},
+} = {}) {
+  if (typeof hidden !== 'function') {
+    warn(
+      `The \`hidden\` option in the "panel" widget expects a function returning a boolean (received "${typeof hidden}" type).`
+    );
+  }
+
+  const cssClasses = {
+    root: cx(suit(), userCssClasses.root),
+    noRefinementRoot: cx(
+      suit({ modifierName: 'noRefinement' }),
+      userCssClasses.noRefinementRoot
+    ),
+    body: cx(suit({ descendantName: 'body' }), userCssClasses.body),
+    header: cx(suit({ descendantName: 'header' }), userCssClasses.header),
+    footer: cx(suit({ descendantName: 'footer' }), userCssClasses.footer),
+  };
+
+  return widgetFactory => (widgetOptions = {}) => {
+    const { container } = widgetOptions;
+
+    if (!container) {
+      throw new Error(
+        `[InstantSearch.js] The \`container\` option is required in the widget "${
+          widgetFactory.name
+        }".`
+      );
+    }
+
+    const defaultTemplates = { header: '', footer: '' };
+    const templateProps = prepareTemplateProps({ defaultTemplates, templates });
+
+    const renderPanel = renderer({
+      containerNode: getContainerNode(container),
+      cssClasses,
+      templateProps,
+    });
+
+    try {
+      const { bodyRef } = renderPanel({
```

Comment:
It feels odd to have to render this outside of the widget init / render lifecycle. It would be nice to always trigger render at this phase, now some elements will start to render before we call `start` on InstantSearch.

Reply:
I see what you mean. The thing is that this kind of circular dependency is necessary to know where to insert the widget.

Another proposal was to render the widget and then the panel, but this comes with a multitude of other problems.